src/nxt_otel.c

```diff
@@ -0,0 +1,403 @@
+
+    nxt_otel_span_add_status(task, r);
+    nxt_otel_state_transition(r->otel, NXT_OTEL_UNINIT_STATE);
+    nxt_otel_rs_send_trace(r->otel->trace);
+
+    r->otel->trace = NULL;
+}
+
+
+void
+nxt_otel_error(nxt_task_t *task, nxt_http_request_t *r)
+{
+    // purposefully not using state transition helper
+    r->otel->status = NXT_OTEL_UNINIT_STATE;
+    nxt_log(task, NXT_LOG_ERR, "otel error condition");
+
+    /* assumable at time of writing that there is no
+     * r->otel->trace to leak. This state is only set
+     * in cases where trace fails to generate or is missing
+     */
+}
+
+
+void
+nxt_otel_trace_and_span_init(nxt_task_t *task, nxt_http_request_t *r)
+{
+    r->otel->trace =
+        nxt_otel_rs_get_or_create_trace(r->otel->trace_id);
+    if (r->otel->trace == NULL) {
+        nxt_log(task, NXT_LOG_ERR, "error generating otel span");
+        nxt_otel_state_transition(r->otel, NXT_OTEL_ERROR_STATE);
+        return;
+    }
+
+    nxt_otel_state_transition(r->otel, NXT_OTEL_HEADER_STATE);
+}
+
+
+void
+nxt_otel_test_and_call_state(nxt_task_t *task, nxt_http_request_t *r)
+{
+    if (r == NULL || r->otel == NULL) {
+        return;
+    }
+
+    switch (r->otel->status) {
+    case NXT_OTEL_UNINIT_STATE:
+      return;
+    case NXT_OTEL_INIT_STATE:
+        nxt_otel_trace_and_span_init(task, r);
+        break;
+    case NXT_OTEL_HEADER_STATE:
+        nxt_otel_span_add_headers(task, r);
+        break;
+    case NXT_OTEL_BODY_STATE:
+        nxt_otel_span_add_body(r);
+        break;
+    case NXT_OTEL_COLLECT_STATE:
+      nxt_otel_span_collect(task, r);
+        break;
+    case NXT_OTEL_ERROR_STATE:
+        nxt_otel_error(task, r);
+        break;
+    }
+}
+
+
+// called in nxt_http_request_error
+void
+nxt_otel_request_error_path(nxt_task_t *task, nxt_http_request_t *r) {
+  if (r->otel->trace == NULL) {
+        return;
+    }
+
+    // response headers have been cleared
+    nxt_otel_propagate_header(task, r);
+
+    // collect span immediately
+    if (r->otel) {
+        nxt_otel_state_transition(r->otel, NXT_OTEL_COLLECT_STATE);
+    }
+    nxt_otel_test_and_call_state(task, r);
+}
+
+
+nxt_int_t
+nxt_otel_parse_traceparent(void *ctx, nxt_http_field_t *field, uintptr_t data)
+{
+    char               *copy;
+    nxt_http_request_t *r;
+
+    /* For information on parsing the traceparent header:
+     * https://www.w3.org/TR/trace-context/#traceparent-header
+     * A summary of the traceparent header value format follows:
+     * Traceparent: "$a-$b-$c-$d"
+     *   a. version (2 hex digits) (ff is forbidden)
+     *   b. trace_id (32 hex digits) (all zeroes forbidden)
+     *   c. parent_id (16 hex digits) (all zeroes forbidden)
+     *   d. flags (2 hex digits)
+     */
+
+    r = ctx;
+    if (field->value_length != NXT_OTEL_TRACEPARENT_LEN) {
+        goto error_state;
+    }
+
+    /* strsep is destructive so we make a copy of the field
+     */
+    copy = nxt_mp_zalloc(r->mem_pool, field->value_length + 1);
```

Comment:
Again, we probably don't need zalloc...

Reply:
I vastly prefer zalloc. Not only do we not need to manually set null terminators but it also makes debugging easier. If you take the step to zero out your memory then later while debugging you can see a segment of memory improperly filled and either not spot a null terminator or see a large series of zeroed data. Also, if it is dereferenced incorrectly you will get more obvious errors.

For what reason is it so superfluous?

Reply:
Sometimes zeroing out memory is desirable sometimes it's not.

If you're using something like sprintf(3) then there's no point as it will add the `\0`, if your copying a nul-terminated string with something like memcpy(3), then again there is no need for the overhead of zeroing out the memory.

Blindly zeroing out memory _may_ also hide uninitialised memory access bugs.